salt/_modules/metalk8s.py

```diff
@@ -87,6 +87,13 @@ def minions_by_role(role, nodes=None):
             Defaults to `pillar.metalk8s.nodes`.
     '''
     nodes = nodes or __pillar__['metalk8s']['nodes']
+    if '_errors' in nodes:
+        log.debug(
+            "Can't retrieve minions by role because of errors in pillar "
+            "'metalk8s:nodes': %s",
+            ', '.join(nodes['_errors']),
+        )
+        return []
```

Comment:
Why you do not raise ? With this we cannot see the difference in output between "no minions with this roles" and "not able to retrieve minions list"

Reply:
If this method is meant to be used directly at render-time, I'd rather have it not fail... But maybe we should just not call it if there are `_errors` in the pillar, and do those checks at every call site.

Reply:
Hmm not sure if something is doable in Salt to catch that kind of raise :thinking: , but what is doable is to return a string if you have an error and a list otherwise so that you just check the result

Reply:
> return a string if you have an error and a list otherwise

Meh, I'm not fond of this kind of trick to handle errors. What I'll do instead, is let this method raise as you suggested, and make sure to not call it if we expect errors:

```sls
{%- if '_errors' in pillar.metalk8s.nodes %}
    {# Bootstrap #}
    {%- set apiservers = ["127.0.0.1"] %}
{%- else %}
    {%- set masters = salt['metalk8s.minions_by_role']('master') %}
    {%- set master_ips = salt['mine.get'](...) %}
    {%- if not master_ips %}
        {{ raise("No master node in pillar!") }}
    {%- endif %}
    {%- set apiservers = master_ips.values() | sort %}
{%- endif %}
```

Reply:
Hmm you sure we have errors in bootstrap ? Because to me we fill the pillar "nodes" properly during bootstrap

Reply:
There's a moment when this gets rendered and there's no kubeconfig yet, so yeah there are some errors in the pillar (I had a failure with something like `None object has no attribute 'items'` from the `minions_by_role` call).

Reply:
Hmm ok weird to me as during bootstrap we override all the pillar contain ...But ok, right

Reply:
Oh right I know what happened! The error wasn't what I shared, but `list item has not attribute 'get'`. It comes for this piece of code:

```python
return [
    node
    for (node, node_info) in nodes.items()
    if role in node_info.get('roles', [])
]
```

The issue: in `metalk8s:nodes`, even if we pass our own value, it'll get merged, so we still have an `_errors` key in the pillar...

Reply:
Guess we can trick it by changing slightly to:

```python
nodes = nodes or __pillar__['metalk8s']['nodes']
pillar_errors = nodes.pop('_errors', None)
if pillar_errors is not None:
    raise CommandExecutionError(
        "Can't retrieve minions by role because of errors in pillar "
        "'metalk8s:nodes': {}".format(', '.join(nodes['_errors']))
    )
```

And override the `_errors` manually when bootstrapping?

Reply:
Yes agree with the last one to me we should override errors when bootstrapping, we fill the pillar so we don't care about ext_pillar errors in these sections

Reply:
Done!